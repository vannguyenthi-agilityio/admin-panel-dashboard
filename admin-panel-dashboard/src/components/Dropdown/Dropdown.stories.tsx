import type { Meta, StoryObj } from '@storybook/react';

// Components
import Dropdown from './';

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction
const meta : Meta<typeof Dropdown> = {
  title: 'Components/Dropdown',
  component: Dropdown,
  tags: ['autodocs'],
  argTypes: {
    user: { description: "Info of user" },
  },
};

export default meta;

type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {},
};

export const NotLogOut: Story = {
  args: {
    hasLogOut: false
  },
};
