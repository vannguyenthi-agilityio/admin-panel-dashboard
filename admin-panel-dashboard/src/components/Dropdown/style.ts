import { cva } from 'class-variance-authority';

export const dropdownStyles = cva(
  'font-primary font-light text-sm text-white border focus:outline-none focus:ring-blue-300 font-medium rounded-lg text-sm px-5 py-2.5 text-center inline-flex items-center',
  {
    variants: {
      variant: {
        primary: ["bg-white", "text-dark"],
        secondary: ["bg-transparent", "text-darker"],
      },
      size: {
        xs: ["text-xs", "px-2", "py-0"],
        sm: ["text-sm", "px-2", "py-1.5"],
        md: ["text-md", "px-2", "py-2"],
        lg: ["text-lg", "px-2", "py-2"],
      },
    },
    defaultVariants: {
      variant: "primary",
      size: "sm",
    },
  }
);
