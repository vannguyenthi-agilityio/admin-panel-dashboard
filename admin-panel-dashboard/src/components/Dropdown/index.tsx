
import { memo, useState, RefObject, useCallback } from "react";

// Hooks
import useOutsideClick from "@/hooks/useOutsideClick";

// Types
import { IDropdown, IListDropDown } from "@/types";

// Helpers
import { clsxMerge } from '@/helpers';

// Icons
import { ArrowDownIcon } from '@/components/Icons';

// Mocks
import { MOCK_USER_DATA , LIST_DROPDOWN_DATA} from '@/mocks';

import { dropdownStyles } from './style';

const Dropdown = (({
  className,
  user = MOCK_USER_DATA,
  listDropDown = LIST_DROPDOWN_DATA,
  hasLogOut = true,
  size,
  ...props
}: IDropdown) => {

  const [isOpenDropdown, setIsOpenDropdown] = useState(false);

  const selectRef = useOutsideClick(() => {
    setIsOpenDropdown(false);
  });

  const handleShowOptionsList = useCallback(
    () => setIsOpenDropdown(prev => !prev),
  []);

  return (
    <div
      ref={selectRef as RefObject<HTMLDivElement>}
      {...props}
    >
      <button
        onClick={handleShowOptionsList}
        data-dropdown-toggle="dropdown"
        type="button"
        className={clsxMerge(
          dropdownStyles({
            size
          }),
          className
        )}
      >
        <img className="w-7 h-7 mr-2 rounded-lg" src={user.avatar}  alt="user photo" />
        <span className="mr-2">{user.name}</span>
        <ArrowDownIcon className="mx-2"/>
      </button>
      {isOpenDropdown &&
        <div id="dropdown" className="z-10 bg-white divide-y divide-gray-100 rounded-lg shadow w-44">
          <ul className="py-2 text-sm text-gray-700 dark:text-gray-200" aria-labelledby="dropdown">
            {listDropDown?.map(
              ({
                name,
                href
              }: IListDropDown) => (
                <li>
                  <a href={href} className="block px-4 py-2 hover:bg-gray-100 dark:hover:bg-gray-600 dark:hover:text-white">{name}</a>
                </li>
              ))
            }
          </ul>
          {hasLogOut &&
            <div className="py-2">
              <a href="#" className="block px-4 py-2 text-sm text-gray-700 hover:bg-gray-100 dark:hover:bg-gray-600 dark:text-gray-200 dark:hover:text-white">Sign out</a>
            </div>
          }
        </div>
      }
    </div>
  )
});

export default memo(Dropdown);
