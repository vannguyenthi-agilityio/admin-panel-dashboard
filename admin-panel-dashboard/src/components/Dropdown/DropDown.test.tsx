import {
  render
} from '@testing-library/react';
import '@testing-library/jest-dom';
import Dropdown from '.'

describe('Dropdown component', () => {
  it('Renders dropdown component', () => {
    const {container} = render(<Dropdown />)
    expect(container).toMatchSnapshot();
  });

  it('Renders dropdown not logout component', () => {
    const {container} = render(<Dropdown hasLogOut={false} />)
    expect(container).toMatchSnapshot();
  });
})
