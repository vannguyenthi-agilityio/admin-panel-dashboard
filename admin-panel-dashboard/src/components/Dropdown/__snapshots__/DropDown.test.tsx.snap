// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown component Renders dropdown component 1`] = `
<div>
  <div>
    <button
      class="font-primary border focus:outline-none focus:ring-blue-300 font-medium rounded-lg text-center inline-flex items-center bg-white text-dark text-sm px-2 py-1.5"
      data-dropdown-toggle="dropdown"
      type="button"
    >
      <img
        alt="user photo"
        class="w-7 h-7 mr-2 rounded-lg"
        src="https://www.gravatar.com/avatar/2c7d99fe281ecd3bcd65ab915bac6dd5?s=250"
      />
      <span
        class="mr-2"
      >
        Derek Alvarado
      </span>
      <svg
        class="mx-2"
        fill="none"
        height="5"
        viewBox="0 0 8 5"
        width="8"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7.25 0.75L4 4.25L0.75 0.75"
          stroke="#adb5bd"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="1.5"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`Dropdown component Renders dropdown not logout component 1`] = `
<div>
  <div>
    <button
      class="font-primary border focus:outline-none focus:ring-blue-300 font-medium rounded-lg text-center inline-flex items-center bg-white text-dark text-sm px-2 py-1.5"
      data-dropdown-toggle="dropdown"
      type="button"
    >
      <img
        alt="user photo"
        class="w-7 h-7 mr-2 rounded-lg"
        src="https://www.gravatar.com/avatar/2c7d99fe281ecd3bcd65ab915bac6dd5?s=250"
      />
      <span
        class="mr-2"
      >
        Derek Alvarado
      </span>
      <svg
        class="mx-2"
        fill="none"
        height="5"
        viewBox="0 0 8 5"
        width="8"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7.25 0.75L4 4.25L0.75 0.75"
          stroke="#adb5bd"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="1.5"
        />
      </svg>
    </button>
  </div>
</div>
`;
